src/watchdog/utils/event_debouncer.py

```diff
@@ -0,0 +1,47 @@
+import logging
+import queue
+
+from watchdog.utils import BaseThread
+
+
+logger = logging.getLogger(__name__)
+
+
+class EventDebouncer(BaseThread):
+    """Background thread for debouncing event handling.
+
+    When an event is received, wait until the configured debounce interval
+    passes before calling the callback.  If additional events are received
+    before the interval passes, reset the timer and keep waiting.  When the
+    debouncing interval passes, the callback will be called with a list of
+    events in the order in which they were received.
+    """
+    def __init__(self, debounce_interval_seconds, events_callback):
+        super().__init__()
+        self.debounce_interval_seconds = debounce_interval_seconds
+        self.events_callback = events_callback
+
+        self._event_queue = queue.Queue()
+
+    def handle_event(self, event):
+        self._event_queue.put(event)
+
+    def run(self):
+        while self.should_keep_running():
+            events = []
+
+            # Get first event.
+            try:
+                events.append(self._event_queue.get(timeout=0.2))
+            except queue.Empty:
+                continue
+
+            if self.debounce_interval_seconds:
+                # Collect additional events until the debounce interval passes.
+                while self.should_keep_running():
+                    try:
+                        events.append(self._event_queue.get(timeout=self.debounce_interval_seconds))
+                    except queue.Empty:
+                        break
+
+            self.events_callback(events)
```

Comment:
Here's a different implementation, like I did it [in MkDocs](https://github.com/mkdocs/mkdocs/blob/56b235a8ad43f2300d17f87e6fa4de7a3d764397/mkdocs/livereload/__init__.py#L166).

It has the benefit of being instantly stoppable, rather than having a 0 - 0.2 sec delay before being able to stop.

It also ensures to instantly stop firing events when it's stopped, maybe  eliminating the need for
> The more involved start and stop mechanisms in AutoRestartTrick are needed to avoid restarting the debouncer when restart the process

```suggestion
import logging
import threading


logger = logging.getLogger(__name__)


class EventDebouncer(threading.Thread):
    """Background thread for debouncing event handling.

    When an event is received, wait until the configured debounce interval
    passes before calling the callback.  If additional events are received
    before the interval passes, reset the timer and keep waiting.  When the
    debouncing interval passes, the callback will be called with a list of
    events in the order in which they were received.
    """
    def __init__(self, debounce_interval_seconds, events_callback):
        super().__init__()
        self.debounce_interval_seconds = debounce_interval_seconds
        self.events_callback = events_callback

        self._cond = threading.Condition()
        self._shutdown = False
        self._events = []

    def handle_event(self, event):
        with self._cond:
            self._events.append(event)
            self._cond.notify_all()

    def run(self):
        while True:
            with self._cond:
                while not self._cond.wait_for(lambda: self._events or self._shutdown, timeout=0.2):
                    # We could have used just one wait instead of a loop + timeout, but we need
                    # occasional breaks, otherwise on Windows we can't receive KeyboardInterrupt.
                    pass
                if self._shutdown:
                    break
                if self.debounce_interval_seconds:
                    while self._cond.wait(timeout=self.debounce_interval_seconds):
                        # Wait for events to stop happening.
                        pass

                events = list(self._events)
                self._events.clear()

            self.events_callback(events)

    def stop(self):
        with self._cond:
            self._shutdown = True
            self._cond.notify_all()
```

Reply:
Thanks for this @oprypin!

Using a condition like this is indeed better for instant stopping, I'll work that into the implementation.

Reply:
@taleinat Well you can just "accept suggestion", I think that will also end up simpler than, for example, keeping it a queue. It's a fully functioning replacement, just in 1 of the new tests will need to add a microscopic sleep to make it pass.

Reply:
Thanks, and your work to make it such a drop-in replacement was helpful and is appreciated!

This version however did not handle well the case where a shutdown signal is given during the debouncing interval. Also, the use of `Condition.wait_for()` is more complex than necessary.

So instead I've integrated the approach while addressing these two concerns.